relay-server/src/actors/global_config.rs

```diff
@@ -0,0 +1,230 @@
+pub struct Get;
+
+/// The message for receiving a watch that subscribes to the [`GlobalConfigService`].
+pub struct Subscribe;
+
+/// A way to get updates of the global config.
+pub enum GlobalConfigManager {
+    /// Returns the most recent global config.
+    Get(Sender<Arc<GlobalConfig>>),
+    /// Returns a [`watch::Receiver`] where global config updates will be sent to.
+    Subscribe(Sender<watch::Receiver<Arc<GlobalConfig>>>),
+}
+
+impl Interface for GlobalConfigManager {}
+
+impl FromMessage<Get> for GlobalConfigManager {
+    type Response = AsyncResponse<Arc<GlobalConfig>>;
+
+    fn from_message(_: Get, sender: Sender<Arc<GlobalConfig>>) -> Self {
+        Self::Get(sender)
+    }
+}
+
+impl FromMessage<Subscribe> for GlobalConfigManager {
+    type Response = AsyncResponse<watch::Receiver<Arc<GlobalConfig>>>;
+
+    fn from_message(_: Subscribe, sender: Sender<watch::Receiver<Arc<GlobalConfig>>>) -> Self {
+        Self::Subscribe(sender)
+    }
+}
+
+/// Service implementing the [`GlobalConfigManager`] interface.
+///
+/// The service offers two alternatives to fetch the [`GlobalConfig`]:
+/// responding to a [`Get`] message with the config for one-off requests, or
+/// subscribing to updates with [`Subscribe`] to keep up-to-date.
+#[derive(Debug)]
+pub struct GlobalConfigService {
+    config: Arc<Config>,
+    /// Sender of the [`watch`] channel for the subscribers of the service.
+    sender: watch::Sender<Arc<GlobalConfig>>,
+    /// Sender of the internal channel to forward global configs from upstream.
+    internal_tx: UnboundedSender<UpstreamResult>,
+    /// Receiver of the internal channel to forward global configs from upstream.
+    internal_rx: UnboundedReceiver<UpstreamResult>,
+    /// Upstream service to request global configs from.
+    upstream: Addr<UpstreamRelay>,
+    /// Handle to avoid multiple outgoing requests.
+    fetch_handle: SleepHandle,
+}
+
+impl GlobalConfigService {
+    /// Creates a new [`GlobalConfigService`].
+    pub fn new(config: Arc<Config>, upstream: Addr<UpstreamRelay>) -> Self {
+        let (sender, _) = watch::channel(Arc::default());
+        let (internal_tx, internal_rx) = mpsc::unbounded_channel();
+        Self {
+            config,
+            sender,
+            internal_tx,
+            internal_rx,
+            upstream,
+            fetch_handle: SleepHandle::idle(),
+        }
+    }
+
+    /// Handles messages from external services.
+    fn handle_message(&self, message: GlobalConfigManager) {
+        match message {
+            GlobalConfigManager::Get(sender) => {
+                sender.send(self.sender.borrow().clone());
+            }
+            GlobalConfigManager::Subscribe(sender) => {
+                sender.send(self.sender.subscribe());
+            }
+        }
+    }
+
+    /// Schedules the next global config request.
+    fn schedule_fetch(&mut self) {
+        if self.fetch_handle.is_idle() {
+            // XXX(iker): set a backoff mechanism?
```

Comment:
i don't really see why we would use a backoff mechanism, isn't that for cases where you'd otherwise have too many frequent retries? if it's just once every 10 seconds, that seems totally fine in my mind